### PR TITLE
Update shipment contents

### PR DESCRIPTION
**Pull request by jrusso1020**
#### What does this PR do?
Allow developers to update a shipments contents through Newgistics API. You can create a new `Newgistics::UpdateShipment` with either/both `add_items` or `remove_items` which are arrays of items to update the shipments contents with. You can do then call `update_shipment.update` on the order to send the changes to Newgistics. If it was successful the call will return true and the success value on the `update_shipment` itself will be set to true.

### CODE REVIEW

File: lib/newgistics/response_handlers/update_shipment_contents.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        if shipment.errors.empty?
+          handle_successful_response(response)
+        end
+      end
+
+      private
+
+      def handle_successful_response(response)
+        xml = Nokogiri::XML(response.body)
+        shipment.success = xml.css('success').children.first.text
```
Comment (manmartinez) [inline lib/newgistics/response_handlers/update_shipment_contents.rb]: I think we can re-write this like:

```ruby
shipment.success = xml.at_css('success').text
```

Also does Virtus parse `"true"` and `"false"` correctly to Booleans?

File: lib/newgistics/update_shipment.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
+module Newgistics
+  class UpdateShipment
```
Comment (manmartinez) [inline lib/newgistics/update_shipment.rb]: I was thinking we could name this as a noun and not a verb to keep it inline with the other models we have, maybe we could name it `Newgistics::ShipmentUpdate`? What do you think?

File: lib/newgistics/update_shipment.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    attribute :success, Boolean
+
+    attribute :errors, Array[String], default: []
+    attribute :warnings, Array[String], default: []
+
+    def success?
+      !!success
+    end
+
+    def update
```
Comment (manmartinez) [inline lib/newgistics/update_shipment.rb]: If we go with the `Newgitics::ShipmentUpdate` name I think we could rename this to `save` to have a similar API with `Newgistics::Order` also I think it fits in nicely:

```ruby
shipment_update = Newgistics::ShipmentUpdate.new(attributes)
shipment_update.save
```

File: lib/newgistics/update_shipment.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      request = Requests::UpdateShipmentContents.new(self)
+      response_handler = ResponseHandlers::UpdateShipmentContents.new(self)
+
+      Newgistics.api.post(request, response_handler)
+
+      success || errors.empty?
```
Comment (manmartinez) [inline lib/newgistics/update_shipment.rb]: I think we should make this: `errors.empty? && success?`. What do you think?

File: spec/cassettes/update_shipment_contents/update/failure.yml [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      string: |
+        <?xml version="1.0"?>
+        <Shipment apiKey="<API_KEY>" id="INVALID">
+          <Items>
+            <Item>
+              <SKU>1007-201-G</SKU>
+              <Qty>1</Qty>
+            </Item>
+          </Items>
+          <AddItems>#&lt;Nokogiri::XML::Builder::NodeBuilder:0x007fb60b18f1c8&gt;</AddItems>
```
Comment (manmartinez) [inline spec/cassettes/update_shipment_contents/update/failure.yml]: This seems weird, why is it logging the `Builder` object as part of the XML?

File: spec/newgistics/update_shipment_spec.rb [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        update_shipment = described_class.new(updated_attributes)
+
+        update_shipment.update
+
+        expect(update_shipment.success).to eq true
+      end
+    end
+
+    vcr_options = { cassette_name: 'update_shipment_contents/update/failure' }
+    context "when the updating the shipment contents fails", vcr: vcr_options do
```
Comment (manmartinez) [inline spec/newgistics/update_shipment_spec.rb]: Should we re-phrase this as: `"when updating the shipment contents fails"`?

File: spec/newgistics/update_shipment_spec.rb [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+      it 'updates the inbound_return object with the errors' do
```
Comment (manmartinez) [inline spec/newgistics/update_shipment_spec.rb]: Should we re-phrase this as `"sets the errors on the shipment update"`?

File: spec/newgistics/update_shipment_spec.rb [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        expect(success).to eq false
+        expect(update_shipment.errors.size).to eq 1
+        expect(update_shipment.errors.first).to eq(
+          "Invalid shipment ID."
+        )
+      end
+    end
+  end
+
+  describe "#success?" do
```
Comment (manmartinez) [inline spec/newgistics/update_shipment_spec.rb]: I think we should de-couple this from the `#update` method, I'd say just test for the return values depending on what the `success` attribute has:

```ruby
it "returns true if success is true"

it "returns false if success is false"

it "returns false if success is nil"
```

File: lib/newgistics/update_shipment.rb [hunk 9]
```diff
@@ -0,0 +1,27 @@
+module Newgistics
+  class UpdateShipment
+    include Virtus.model
+
+    attribute :id, String
+    attribute :order_id, String
+    attribute :add_items, Array[Item]
```
Comment (manmartinez) [inline lib/newgistics/update_shipment.rb]: Do we need to set a default value for this attribute? Maybe default to an empty array like we do with errors and warnings?

File: lib/newgistics/response_handlers/update_shipment_contents.rb (hunk 1 above)
Comment (jrusso1020) [reply lib/newgistics/response_handlers/update_shipment_contents.rb]: yes Virtus will correctly parse `"true"` and `"yes"` to `true`and `"false"` and `"no"` to `false`

File: spec/cassettes/update_shipment_contents/update/failure.yml (hunk 5 above)
Comment (jrusso1020) [reply spec/cassettes/update_shipment_contents/update/failure.yml]: this was an error on my part in the code, thanks for catching!

File: spec/newgistics/shipment_update_spec.rb [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+  describe "#success?" do
+    before { use_valid_api_key }
+
+    it "returns false if success is nil" do
+      update_shipment = described_class.new
+
+      expect(update_shipment.success?).to eq false
+    end
+
+    it 'returns true if success is false' do
```
Comment (manmartinez) [inline spec/newgistics/shipment_update_spec.rb]: Can we re-phrase this as `"returns true if success is true"`?

File: spec/newgistics/shipment_update_spec.rb [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+      update_shipment = described_class.new
+
+      update_shipment.success = true
+
+      expect(update_shipment.success?).to eq true
+    end
+
+    it 'returns false' do
```
Comment (manmartinez) [inline spec/newgistics/shipment_update_spec.rb]: Can we add why it returns `false`? Maybe `"returns false if success is false"`?

**Pull request merged**